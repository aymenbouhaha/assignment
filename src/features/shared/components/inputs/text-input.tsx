import { cn } from "@/lib/utils";
import { Input } from "@shared/components/ui";
import { CustomIcon, Icon } from "@shared/components/icons/icons.tsx";
import { useEffect, useState } from "react";

interface TextInputProps {
	placeholder: string;
	leadIcon?: {
		icon: CustomIcon;
		className?: string;
	};
	onValueChange?: (value: string) => void;
	inputClassName?: string;
}

export const TextInput = ({ placeholder, inputClassName, leadIcon, onValueChange }: TextInputProps) => {
	const [value, setValue] = useState<string>("");

	useEffect(() => {
		if (onValueChange) onValueChange(value);
	}, [value]);

	return (
		<div className="w-full relative">
			{leadIcon && (
				<div className="absolute inset-y-0 left-3 flex items-center">
					{Icon[leadIcon.icon]({ className: cn("size-6", leadIcon.className) })}
				</div>
			)}
			<Input
				placeholder={placeholder}
				value={value}
				onChange={(e) => setValue(e.target.value)}
				className={cn(inputClassName, {
					"pl-10": leadIcon,
				})}
			/>
			{value && (
				<Icon.Close
					className={"stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"}
					onClick={() => setValue("")}
				/>
			)}
		</div>
	);
};
