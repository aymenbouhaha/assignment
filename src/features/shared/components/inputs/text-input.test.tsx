import { render, fireEvent } from "@testing-library/react";
import { describe, it, expect } from "vitest";

import { vi } from "vitest";
import { TextInput } from "@shared/components/inputs/text-input.tsx";

describe("TextInput", () => {
	it("matches the snapshot", () => {
		const { container } = render(<TextInput placeholder="Enter text" />);
		expect(container).toMatchSnapshot();
	});

	it("renders lead icon when provided", () => {
		const { container } = render(<TextInput placeholder="Enter text" leadIcon={{ icon: "Search" }} />);
		expect(container).toMatchSnapshot();
	});

	it("calls onValueChange when input value changes", () => {
		const mockOnValueChange = vi.fn();
		const { getByPlaceholderText } = render(<TextInput placeholder="Enter text" onValueChange={mockOnValueChange} />);

		const input = getByPlaceholderText("Enter text");
		fireEvent.change(input, { target: { value: "Hello" } });

		expect(mockOnValueChange).toHaveBeenCalledWith("Hello");
	});

	it("clears the input value when close icon is clicked", () => {
		const { getByPlaceholderText, container } = render(<TextInput placeholder="Enter text" />);

		const input = getByPlaceholderText("Enter text");
		fireEvent.change(input, { target: { value: "Hello" } });

		expect(input).toHaveValue("Hello");

		const closeIcon = container.querySelector(".cursor-pointer");
		fireEvent.click(closeIcon!);

		expect(input).toHaveValue("");
	});
});
