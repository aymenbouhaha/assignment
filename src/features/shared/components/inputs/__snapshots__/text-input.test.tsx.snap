// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInput > matches the snapshot 1`] = `
<div>
  <div
    class="w-full relative"
  >
    <input
      class="file:bg-transparent file:text-sm file:text-zinc-950 flex w-full rounded-md border border-grey-border px-3 py-2.5 text-p21 lg:text-p3 ring-0 outline-none focus:border-grey-2 file:border-0 file:font-medium placeholder:text-grey-1 disabled:cursor-not-allowed disabled:opacity-50"
      placeholder="Enter text"
      value=""
    />
  </div>
</div>
`;

exports[`TextInput > renders lead icon when provided 1`] = `
<div>
  <div
    class="w-full relative"
  >
    <div
      class="absolute inset-y-0 left-3 flex items-center"
    >
      <svg
        class="size-6"
        fill="none"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.7888 11.5843C8.8011 13.1763 5.89152 13.051 4.04893 11.2084C2.0719 9.23138 2.0719 6.02598 4.04893 4.04895C6.02596 2.07192 9.23136 2.07192 11.2084 4.04895C13.051 5.89154 13.1763 8.8011 11.5843 10.7888L15.451 14.6555C15.6707 14.8752 15.6707 15.2314 15.451 15.451C15.2313 15.6707 14.8752 15.6707 14.6555 15.451L10.7888 11.5843ZM4.84443 10.4129C3.30674 8.87522 3.30674 6.38213 4.84443 4.84444C6.38212 3.30675 8.87521 3.30676 10.4129 4.84444C11.9495 6.38101 11.9506 8.87156 10.4163 10.4095C10.4151 10.4106 10.414 10.4118 10.4129 10.4129C10.4118 10.414 10.4106 10.4152 10.4095 10.4163C8.87155 11.9506 6.38099 11.9495 4.84443 10.4129Z"
          fill="#72767C"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <input
      class="file:bg-transparent file:text-sm file:text-zinc-950 flex w-full rounded-md border border-grey-border px-3 py-2.5 text-p21 lg:text-p3 ring-0 outline-none focus:border-grey-2 file:border-0 file:font-medium placeholder:text-grey-1 disabled:cursor-not-allowed disabled:opacity-50 pl-10"
      placeholder="Enter text"
      value=""
    />
  </div>
</div>
`;
